### DISCUSSION TITLE
Output result for RPC Download in bbf-software-management.yang

### DISCUSSION THREAD

**Discussion by inangovshirinkin** [CESNET/libyang#2145] [Q&A] [answered]
I am has problem with find output for RPC Download via function libyang::lys_find_xpath() ...

1. Call with LYS_FIND_XP_OUTPUT didn't found anything
<pre>
lys_find_xpath(ctx, NULL, "/ietf-hardware:hardware/component[name='switch']/bbf-software-management:software/software[name='cpu_sw']/revisions/download/download/*", LYS_FIND_XP_OUTPUT, &items)
</pre>
<details>
<pre>
[T 63] lys_find_xpath(545)Begin: ctx=0x55fabfb0, ctx_node=0, xpath='/ietf-hardware:hardware/component[name='dpu_switch']/bbf-software-management:software/software[name='banner']/revisions/download/download/*', options=16
[T 63] lys_find_xpath(561)Allocate return set ...
[T 63] lys_find_xpath(566)Compile expression ...
[T 63] lys_find_xpath(569)Exp parsed: used=26, size=30
[T 63] lys_find_xpath(572)Atomize expression ...
[T 63] lys_find_xpath(577)Transform into ly_set ...
[T 63] lys_find_xpath(582)Iterations up to 4 ...
[T 63] lys_find_xpath(584).. [0] type=2, in_ctx=0
[T 63] lys_find_xpath(584).. [1] type=3, in_ctx=0
[T 63] lys_find_xpath(586).. [1] name='hardware'
[T 63] lys_find_xpath(584).. [2] type=3, in_ctx=0
[T 63] lys_find_xpath(586).. [2] name='component'
[T 63] lys_find_xpath(587).. [2] ref='RFC 6933: Entity MIB (Version 4) - entPhysicalEntry'
[T 63] lys_find_xpath(584).. [3] type=3, in_ctx=1
[T 63] lys_find_xpath(586).. [3] name='name'
[T 63] lys_find_xpath(597)Cleanup ...
[T 63] lys_find_xpath(604)Return: 0
</pre>
</details>

2. Calling without any option was returned RPC Download input nodes
<pre>
lys_find_xpath(ctx, NULL, "/ietf-hardware:hardware/component[name='switch']/bbf-software-management:software/software[name='cpu_sw']/revisions/download/download/*", 0, &items)
</pre>
<details>
<pre>
[T 63] lys_find_xpath(545)Begin: ctx=0x55ff6fb0, ctx_node=0, xpath='/ietf-hardware:hardware/component[name='switch']/bbf-software-management:software/software[name='cpu_sw']/revisions/download/download/*', options=0
[T 63] lys_find_xpath(561)Allocate return set ...
[T 63] lys_find_xpath(566)Compile expression ...
[T 63] lys_find_xpath(569)Exp parsed: used=26, size=30
[T 63] lys_find_xpath(572)Atomize expression ...
[T 63] lyxp_atomize(9912)Begin
[T 63] lyxp_atomize(9944)Return: 0
[T 63] lys_find_xpath(577)Transform into ly_set ...
[T 63] lys_find_xpath(582)Iterations up to 12 ...
[T 63] lys_find_xpath(584).. [0] type=1, in_ctx=0
[T 63] lys_find_xpath(584).. [1] type=3, in_ctx=0
[T 63] lys_find_xpath(586).. [1] name='hardware'
[T 63] lys_find_xpath(584).. [2] type=3, in_ctx=0
[T 63] lys_find_xpath(586).. [2] name='component'
[T 63] lys_find_xpath(587).. [2] ref='RFC 6933: Entity MIB (Version 4) - entPhysicalEntry'
[T 63] lys_find_xpath(584).. [3] type=3, in_ctx=1
[T 63] lys_find_xpath(586).. [3] name='name'
[T 63] lys_find_xpath(584).. [4] type=3, in_ctx=0
[T 63] lys_find_xpath(586).. [4] name='software'
[T 63] lys_find_xpath(584).. [5] type=3, in_ctx=0
[T 63] lys_find_xpath(586).. [5] name='software'
[T 63] lys_find_xpath(584).. [6] type=3, in_ctx=1
[T 63] lys_find_xpath(586).. [6] name='name'
[T 63] lys_find_xpath(584).. [7] type=3, in_ctx=0
[T 63] lys_find_xpath(586).. [7] name='revisions'
[T 63] lys_find_xpath(584).. [8] type=3, in_ctx=0
[T 63] lys_find_xpath(586).. [8] name='download'
[T 63] lys_find_xpath(584).. [9] type=3, in_ctx=0
[T 63] lys_find_xpath(586).. [9] name='download'
[T 63] lys_find_xpath(584).. [10] type=3, in_ctx=2
[T 63] lys_find_xpath(586).. [10] name='source'
[T 63] lys_find_xpath(591).. [10] Was added
[T 63] lys_find_xpath(584).. [11] type=3, in_ctx=2
[T 63] lys_find_xpath(586).. [11] name='target'
[T 63] lys_find_xpath(591).. [11] Was added
[T 63] lys_find_xpath(597)Cleanup ...
[T 63] lys_find_xpath(604)Return: 0
</pre>
</details>

What am I doing wrong?
And how find output nodes for RPC Download?

**Comment by michalvasko** [comment, marked as answer]
Please try the latest *libyang* in `devel`, there was a bug that should be fixed.

**Comment by inangovshirinkin** [comment]
Thank you for answer! Will check fix from devel

**Comment by inangovshirinkin** [comment]
I'm was ckeck with commit 5357561 from  devel<br>
but outputs not worked good

**Reply by michalvasko** [reply]
I have [tested](https://github.com/CESNET/libyang/files/13703509/test.c.txt) it with the current `devel` and it worked fine. If you have only taken the bugfix, try using the actual current version of *libyang*. If it still does not work, provide a similar example to mine to reproduce the problem.

**Comment by inangovshirinkin** [comment]
<pre>
$ ./test 
Found with output:
Found w/o output:
	source
	target
</pre>

libyang build from last devel
<pre>
$ git log
commit eecad552fe58326a646df9cc9d7fea145ed64e3b (HEAD -> devel, origin/devel)
Author: Michal Vasko <mvasko@cesnet.cz>
Date:   Fri Dec 15 16:09:57 2023 +0100

    VERSION bump to version 2.1.144
</pre>

**Reply by inangovshirinkin** [reply]
<pre>
$ sudo make install
...
-- Up-to-date: /usr/local/lib/libyang.so.2.45.8
-- Up-to-date: /usr/local/lib/libyang.so.2
-- Up-to-date: /usr/local/lib/libyang.so
...
$ LD_LIBRARY_PATH=/usr/local/lib/ ./test
Found with output:
	target
Found w/o output:
	source
	target
$
</pre>

**Reply by michalvasko** [reply]
Okay, so it works for you now.

**Reply by inangovshirinkin** [reply]
shouldn't it return the `target/id` ?

**Reply by michalvasko** [reply]
No, `*` returns only the children nodes. If you want the descendants of those, recursively, as well, you can use `*//.`, for example.

**Reply by inangovshirinkin** [reply]
Thank you for answers!

**Comment by michalvasko** [comment]
Please provide similar output.
```
> ./test 
Found with output:
        target
Found w/o output:
        source
        target
> ldd ./test
        linux-vdso.so.1 (0x00007ffe597f7000)
        libyang.so.2 => /usr/local/lib64/libyang.so.2 (0x00007f7e3076d000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f7e30400000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f7e30686000)
        libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007f7e30359000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f7e3090c000)
> ls -l /usr/local/lib64/libyang.so*
lrwxrwxrwx 1 root root      12 dec 18 14:53 /usr/local/lib64/libyang.so -> libyang.so.2
lrwxrwxrwx 1 root root      17 dec 18 14:53 /usr/local/lib64/libyang.so.2 -> libyang.so.2.45.8
-rwxr-xr-x 1 root root 4022248 dec 15 16:10 /usr/local/lib64/libyang.so.2.45.8
```
